### PR TITLE
Handle reorgs for agent

### PR DESCRIPTION
**Pull request by bilbeyt**
Closes: #1570

### CODE REVIEW

File: beamer/agent/chain.py
Comment (istankovic) [inline beamer/agent/chain.py:29]: Could you please make this configurable, per-chain, similar to what Manuel has done for the poll period?
We should have one global default which is used in case the chain-specific value is not defined.

File: beamer/agent/events.py
Comment (istankovic) [inline beamer/agent/events.py:327]: I feel like the entire block above could be simplified by just subtracting the number of confirmation blocks from `block_number` after getting the block number from the RPC, i.e. after that `try...except` block above. Then you would not need to change anything here, right?

File: beamer/agent/config.py
Comment (istankovic) [inline beamer/agent/config.py:122]: I'm guessing this `int(...)` here is only because of mypy... if so, :woozy_face:

File: beamer/agent/events.py
Comment (istankovic) [inline beamer/agent/events.py:321]: Why do we need to cast `to_block` to `BlockNumber`? It was without it before...

File: beamer/agent/events.py
Comment (istankovic) [inline beamer/agent/events.py:308]: Nitpick: this is totally fine, but I'd prefer to have an `assert block_number > 0` after this anyway...
Also, only the `get_block` call should be inside of this `try...except` block, the rest does not need to be there so let's move it out of the block.

File: beamer/agent/config.py
Comment (istankovic) [inline beamer/agent/config.py:27]: Nitpick: just `confirmation_blocks` would be shorter and a bit nicer....

File: beamer/tests/agent/test_request.py
Comment (istankovic) [inline beamer/tests/agent/test_request.py:397]: How does this test work, i.e. how does a reorg happen and how do we check that we handled it correctly?

Comment (bilbeyt) [reply beamer/tests/agent/test_request.py:397]: so basically without extra ape.chain.mine(1) there we dont get the request at all, it is working in this way.

Comment (bilbeyt) [reply beamer/tests/agent/test_request.py:397]: so if request in block number 23 normally we can only reach it in block number 24 iterationn which we want

File: beamer/agent/events.py
Comment (bilbeyt) [reply beamer/agent/events.py:321]: my bad :D forgot it

File: docs/source/commands.rst
Comment (istankovic) [inline docs/source/commands.rst:36]: Nitpick: `block is ready to process` -> `block ready for processing`

File: docs/source/configuration.rst
Comment (istankovic) [inline docs/source/configuration.rst:159]: Nitpick: `block is ready to process` -> `block ready for processing`

File: beamer/tests/agent/test_request.py
Comment (istankovic) [reply beamer/tests/agent/test_request.py:397]: I see, but the problem is, this test would be successful even without the confirmation blocks feature, right?. I.e. we are not handling a reorg _really_ because there is no reorg. We are just testing whether we wait for the specified number of confirmation blocks.

Comment (bilbeyt) [reply beamer/tests/agent/test_request.py:397]: isnt it enough? can you do a reorg in test?

Comment (istankovic) [reply beamer/tests/agent/test_request.py:397]: Well, one way to do it would be to 
1. make a snapshot
2. mine N blocks
3. roll back the chain to the snapshot
4. mine new blocks

I think that's as close as we could get to a real reorg with ape.
Now, for the actual test...

1. make a snapshot
2. mine N blocks, where `N > confirmation blocks`
3. make a request in another block, note the request ID, say `request1`
4. roll back the chain to the snapshot
5. mine N blocks
6. assert that the agent _did not_ see `request1`
7. make another request, note the request ID, say `request2`
8. mine `confirmation blocks` new blocks
9. assert that the agent _did_ see `request2`

**Review by istankovic — CHANGES_REQUESTED** [review]
Looks good overall,  but I think we need to improve the test.

File: beamer/agent/data/agent.conf.example
Comment (fredo) [inline beamer/agent/data/agent.conf.example:5]: What is the meaning of the confirmation-blocks on the top level?

File: beamer/agent/config.py
Comment (fredo) [inline beamer/agent/config.py:120]: general suggestion: Wouldn't it be nicer to have a "chain" dictionary in the config which stores chain related information. 
Since we sort our information per chain we could have a dataclass `ChainConfig`
and a `dict[chain_name, ChainConfig]` (kind of similar how we structured it with api schema in the deployment config

File: beamer/agent/data/agent.conf.example
Comment (fredo) [reply beamer/agent/data/agent.conf.example:5]: is it meant to be a default?

File: beamer/agent/events.py
Comment (bilbeyt) [reply beamer/agent/events.py:308]: not sure about assert here actually, lets say confirmation blocks 1 and block number is 1 then this makes 0 and assert will crash, i would have it as if and return [] instead so this wont crash

Comment (bilbeyt) [reply beamer/agent/events.py:308]: not adding any assert would be doing it actually

File: beamer/agent/data/agent.conf.example
Comment (bilbeyt) [reply beamer/agent/data/agent.conf.example:5]: yes it will be default. the first suggestion was using the structure like we are using for poll_period and chain based poll_period so i have added also this one as default.

File: beamer/agent/config.py
Comment (bilbeyt) [reply beamer/agent/config.py:120]: i can handle this after merging this pr

Comment (istankovic) [reply beamer/agent/config.py:120]: > and a `dict[chain_name, ChainConfig]` (kind of similar how we structured it with api schema in the deployment config

Exactly this has been my plan to do, just didn't want to block this PR on that. But yes, using the apischema approach for agent config is definitely the way to go.

File: beamer/agent/events.py
Comment (istankovic) [reply beamer/agent/events.py:308]: > not adding any assert would be doing it actually

I'm afraid I don't understand this. :sweat_smile:

Comment (bilbeyt) [reply beamer/agent/events.py]: so block_number < 0 will return [] in next if at line 317 because self._next_block_number would be bigger

Comment (istankovic) [reply beamer/agent/events.py]: > so block_number < 0 will return [] in next if at line 317 because self._next_block_number would be bigger

Ah, yes, perfect! :+1:

File: beamer/tests/agent/test_request.py
Comment (istankovic) [inline beamer/tests/agent/test_request.py:389]: Nitpick: I would just call this `snapshot` instead of `snap1`.

File: beamer/tests/agent/test_request.py
Comment (istankovic) [inline beamer/tests/agent/test_request.py:392]: Here it could happen that the agent does not get a chance to sync before we restore the chain state.
So before doing `restore`, we should wait until event fetcher's `synced_block` is `latest chain's block number - confirmation block`. Only then we should do a `restore`.

File: beamer/agent/config.py
Comment (fredo) [reply beamer/agent/config.py]: ok then lets go for it after this PR, can you create a follow up please?

File: beamer/tests/agent/test_request.py
Comment (istankovic) [inline beamer/tests/agent/test_request.py:393]: We're checking whether the block number is zero? Is that intentional?

File: beamer/agent/data/agent.conf.example
Comment (fredo) [reply beamer/agent/data/agent.conf.example:6]: Just to clarify: 
The agent has the option to configure the global default, plus for chain specific values, which override the default? 
Additionally, we have an internal default which is used if no global default is set? 
And this counts both for poll period and confirmation blocks?

File: beamer/tests/agent/test_request.py
Comment (istankovic) [inline beamer/tests/agent/test_request.py:400]: Very nice solution, that is exactly what we needed!

File: beamer/agent/data/agent.conf.example
Comment (fredo) [reply beamer/agent/data/agent.conf.example:6]: (if true, sounds to me like an overkill)

File: beamer/tests/agent/test_request.py
Comment (bilbeyt) [reply beamer/tests/agent/test_request.py:393]: it is mypy again and the reason is actually this expression return type it is int | None so i am making sure that this is not None

File: beamer/agent/data/agent.conf.example
Comment (istankovic) [reply beamer/agent/data/agent.conf.example:6]: > Just to clarify: The agent has the option to configure the global default, plus for chain specific values, which override the default? Additionally, we have an internal default which is used if no global default is set? And this counts both for poll period and confirmation blocks?

Yes, exactly.

I do admit it may seem as a bit of an overkill and most likely not needed (at least not right now), but OTOH, having it in this manner is more in line with how people generally expect configuration to behave, so I see it not as "we're definitely not going to need this" but rather "people expect config files to work a certain way, so let's make it intuitive for them, even though they may not have an immediate need for it".

File: beamer/tests/agent/test_request.py
Comment (istankovic) [reply beamer/tests/agent/test_request.py:393]: > it is mypy again and the reason is actually this expression return type it is int | None so i am making sure that this is not None

Ah, mypy again. :woozy_face: 
But then, please make it explicit that you are testing against `None`:

```python
    assert ape.chain.blocks[-1].number is not None
```

**Review by istankovic — APPROVED** [review]
Looks good from my side.

@fredo may have some reservations about this, though.

**Pull request merged**
